@johnynek here is a test which fails during reduce phase with exception `Caused by: java.lang.ClassCastException: com.twitter.scalding.OrderedSerializationAndDistinctJob$User cannot be cast to com.twitter.scalding.OrderedSerializationAndDistinctJob$Student`:
```
object OrderedSerializationAndDistinctJob {
  class User(val id: Long)

  implicit val ord: OrderedSerialization[User] =
    OrderedSerialization.viaTransform[User, Long](_.id, new User(_))(RequiredBinaryComparators.orderedSerialization[Long])

  class Student(override val id: Long, val name: String) extends User(id)
}

class OrderedSerializationAndDistinctJob(args: Args) extends Job(args) {
  import OrderedSerializationAndDistinctJob._

  TypedPipe
    .from(0 to 100)
    .map(id => new Student(id, "name"))
    .distinct
    .map(_.name)
    .write(TypedText.tsv[String]("output"))
}

class OrderedSerializationAndDistinctJobTest extends WordSpec with Matchers {
  "Distinct should work" should {
    JobTest(new OrderedSerializationAndDistinctJob(_))
      .sink[String](TypedText.tsv[String]("output")) { outputBuffer =>
        val data = outputBuffer.toList
        assert(data.size == 100, "should be 100 elements")
        assert(data.forall { _ == "name" }, "all elements should be name")
      }
      .runHadoop.finish()
  }
}
```

I'm not putting this test as part of PR because with this PR this test doesn't compile.
I know it's kinda artificial use case because I've implemented `OrderedSerialization` on my own which people don't do, so I'm not really picking on this PR. But issue is that `OrderedSerialization` is conceptually invariant - it also serialize and deserialize things.

I think my biggest point here - it was literally 6 places over twitter repo to fix it so I don't think it actually gonna add pain for upgrades while it can lead to code which compiles, don't have `asInstanceOf` and still fails at runtime.

--- a/scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala
+++ b/scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala
@@ -207,7 +207,25 @@ object TypedPipe extends Serializable {
    }
 
    implicit class InvariantTypedPipe[T](val pipe: TypedPipe[T]) extends AnyVal {
-      /**
+     /**
+      * Returns the set of distinct elements in the TypedPipe
+      * This is the same as: .map((_, ())).group.sum.keys
+      * If you want a distinct while joining, consider:
+      * instead of:
+      * {@code
+      *   a.join(b.distinct.asKeys)
+      * }
+      * manually do the distinct:
+      * {@code
+      *   a.join(b.asKeys.sum)
+      * }
+      * The latter creates 1 map/reduce phase rather than 2
+      */
+     @annotation.implicitNotFound(msg = "For distinct method to work, the type in TypedPipe must have an Ordering.")
+     def distinct(implicit ord: Ordering[T]): TypedPipe[T] =
+       pipe.asKeys.sum.keys
+
+     /**
        * If any errors happen below this line, but before a groupBy, write to a TypedSink
        */
       def addTrap(trapSink: Source with TypedSink[T])(implicit conv: TupleConverter[T]): TypedPipe[T] =
@@ -631,24 +649,6 @@ sealed abstract class TypedPipe[+T] extends Serializable with Product {
   /** adds a description to the pipe */
   def withDescription(description: String): TypedPipe[T] =
     TypedPipe.WithDescriptionTypedPipe[T](this, (description, false) :: Nil)
-
-  /**
-   * Returns the set of distinct elements in the TypedPipe
-   * This is the same as: .map((_, ())).group.sum.keys
-   * If you want a distinct while joining, consider:
-   * instead of:
-   * {@code
-   *   a.join(b.distinct.asKeys)
-   * }
-   * manually do the distinct:
-   * {@code
-   *   a.join(b.asKeys.sum)
-   * }
-   * The latter creates 1 map/reduce phase rather than 2
-   */
-  @annotation.implicitNotFound(msg = "For distinct method to work, the type in TypedPipe must have an Ordering.")
-  def distinct[U >: T](implicit ord: Ordering[U]): TypedPipe[U] =
-    asKeys[U].sum.keys
 
   /**
    * Returns the set of distinct elements identified by a given lambda extractor in the TypedPipe

